witchcraft-server/src/service/tls.rs

```diff
@@ -15,158 +15,176 @@ use crate::service::hyper::NewConnection;
 ];
 
-const TLS13_CIPHER_SUITES: &[&str] = &[
-    "TLS_AES_256_GCM_SHA384",
-    "TLS_AES_128_GCM_SHA256",
-    "TLS_CHACHA20_POLY1305_SHA256",
-];
+static KX_GROUPS: [&SupportedKxGroup; 3] = [&SECP256R1, &SECP384R1, &X25519];
+
+static PROTOCOL_VERSIONS: [&SupportedProtocolVersion; 2] = [&TLS12, &TLS13];
 
 /// A layer which wraps streams in a TLS session.
 pub struct TlsLayer {
-    context: SslContext,
+    acceptor: TlsAcceptor,
 }
 
 impl TlsLayer {
     pub fn new(config: &InstallConfig) -> Result<Self, Error> {
-        let mut builder = SslContext::builder(SslMethod::tls()).map_err(Error::internal_safe)?;
-
-        builder
-            .set_certificate_chain_file(config.keystore().cert_path())
-            .map_err(Error::internal_safe)?;
-        builder
-            .set_private_key_file(config.keystore().key_path(), SslFiletype::PEM)
+        let builder = ServerConfig::builder()
+            .with_cipher_suites(&CIPHER_SUITES)
+            .with_kx_groups(&KX_GROUPS)
+            .with_protocol_versions(&PROTOCOL_VERSIONS)
             .map_err(Error::internal_safe)?;
-        builder.check_private_key().map_err(Error::internal_safe)?;
 
-        if let Some(client_auth) = config.client_auth_truststore() {
-            builder
-                .set_ca_file(client_auth.path())
-                .map_err(Error::internal_safe)?;
+        let builder = match config.client_auth_truststore() {
+            Some(client_auth_truststore) => {
+                let certs = load_certificates(client_auth_truststore.path())?;
+                let mut store = RootCertStore::empty();
+                store.add_parsable_certificates(&certs);
+                builder.with_client_cert_verifier(Arc::new(
+                    AllowAnyAnonymousOrAuthenticatedClient::new(store),
+                ))
+            }
+            None => builder.with_no_client_auth(),
+        };
 
-            builder.set_verify(SslVerifyMode::PEER);
-        }
+        let cert_chain = load_certificates(config.keystore().cert_path())?
+            .into_iter()
+            .map(Certificate)
+            .collect();
 
-        let cipher_list = convert_suites(CIPHER_SUITES);
-        builder
-            .set_cipher_list(&cipher_list)
-            .map_err(Error::internal_safe)?;
+        let key_der = load_private_key(config.keystore().key_path())?;
 
-        let ciphersuites = convert_suites(TLS13_CIPHER_SUITES);
-        builder
-            .set_ciphersuites(&ciphersuites)
+        let mut server_config = builder
+            .with_single_cert(cert_chain, key_der)
             .map_err(Error::internal_safe)?;
 
+        server_config.ignore_client_order = true;
         if config.server().http2() {
-            builder.set_alpn_select_callback(|_, client| {
-                ssl::select_next_proto(b"\x02h2\x08http/1.1", client).ok_or(AlpnError::ALERT_FATAL)
-            });
+            server_config.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
```

Comment:
nit: Worth making the available protocols a static constant?

Reply:
It's only used in this one place, so I don't really think it's worth it.